src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs

```diff
@@ -22,56 +21,71 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.C
+        {
+            return _threadingContext.JoinableTaskFactory.Run(() =>
+                this.CreateProjectContextAsync(languageName, projectUniqueName, projectFilePath, projectGuid, hierarchy, binOutputPath, CancellationToken.None));
+        }
+
+        public async Task<IWorkspaceProjectContext> CreateProjectContextAsync(
             string languageName,
             string projectUniqueName,
             string projectFilePath,
             Guid projectGuid,
             object hierarchy,
-            string binOutputPath)
+            string binOutputPath,
+            CancellationToken cancellationToken)
         {
-            var visualStudioProject = CreateVisualStudioProject(languageName, projectUniqueName, projectFilePath, hierarchy as IVsHierarchy, projectGuid);
-            return new CPSProject(visualStudioProject, _workspace, _projectCodeModelFactory, projectGuid, binOutputPath);
-        }
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-        private VisualStudioProject CreateVisualStudioProject(string languageName, string projectUniqueName, string projectFilePath, IVsHierarchy hierarchy, Guid projectGuid)
-        {
             var creationInfo = new VisualStudioProjectCreationInfo
             {
                 FilePath = projectFilePath,
-                Hierarchy = hierarchy,
+                Hierarchy = hierarchy as IVsHierarchy,
                 ProjectGuid = projectGuid,
             };
 
-            var visualStudioProject = _projectFactory.CreateAndAddToWorkspace(projectUniqueName, languageName, creationInfo);
+            var visualStudioProject = await _projectFactory.CreateAndAddToWorkspaceAsync(
+                projectUniqueName, languageName, creationInfo, cancellationToken).ConfigureAwait(true);
+
+            // At this point we've mutated the workspace.  So we're no longer cancellable.
+            cancellationToken = CancellationToken.None;
 
             if (languageName == LanguageNames.FSharp)
             {
-                var shell = (IVsShell)ServiceProvider.GlobalProvider.GetService(typeof(SVsShell));
+                var shell = await _serviceProvider.GetServiceAsync<SVsShell, IVsShell7>().ConfigureAwait(true);
 
                 // Force the F# package to load; this is necessary because the F# package listens to WorkspaceChanged to 
                 // set up some items, and the F# project system doesn't guarantee that the F# package has been loaded itself
                 // so we're caught in the middle doing this.
-                shell.LoadPackage(Guids.FSharpPackageId, out _);
+                var packageId = Guids.FSharpPackageId;
+                await shell.LoadPackageAsync(ref packageId);
             }
 
-            return visualStudioProject;
+            // CPSProject constructor has a UI thread dependencies currently, so switch back to the UI thread before proceeding.
```

Comment:
```suggestion
            // CPSProject constructor has a UI thread dependencies currently, so we have to stay on the UI thread from the earlier switch.
```

Odd there wasn't an active switch happening here.